binder/src/main/java/io/grpc/binder/internal/BinderClientTransport.java

```diff
@@ -391,6 +401,63 @@ private synchronized void handleAuthResult(Throwable t) {
         Status.INTERNAL.withDescription("Could not evaluate SecurityPolicy").withCause(t), true);
   }
 
+  /**
+   * Updates in-use stream count and notifies listener only on transitions between 0 and >0, without
+   * acquiring the transport lock.
+   */
+  private void updateInUseStreamsCountIfNeeded(boolean countsForInUse, int delta) {
+    Preconditions.checkArgument(delta == -1 || delta == 1, "stream count delta must be -1 or +1");
+    if (!countsForInUse) {
+      return;
+    }
+    int prev, next;
+
+    if (delta > 0) {
+      next = numInUseStreams.incrementAndGet();
+      prev = next - 1;
+    } else {
+      prev = numInUseStreams.get();
+      int updated;
+
+      while (true) {
+        int current = prev;
+        int newValue = current > 0 ? current - 1 : 0;
+        if (numInUseStreams.compareAndSet(current, newValue)) {
+          updated = newValue;
+          break;
+        }
+        prev = numInUseStreams.get();
+      }
+      next = updated;
+    }
+
+    boolean prevInUse = prev > 0;
+    boolean nextInUse = next > 0;
+
+    if (prevInUse != nextInUse) {
+      if (listenerInUse.compareAndSet(prevInUse, nextInUse)) {
+        scheduleTransportInUseNotification(nextInUse);
+      }
+    }
+  }
+
+  private void scheduleTransportInUseNotification(final boolean inUse) {
+    getScheduledExecutorService()
+        .execute(
+            new Runnable() {
+              @Override
+              public void run() {
+                // Provide external synchronization as required by Listener contract,
+                // without taking the transport lock to avoid potential deadlocks.
+                synchronized (listenerNotifyLock) {
+                  if (listenerInUse.get() == inUse) {
+                    clientTransportListener.transportInUse(inUse);
```

Comment:
Your new `listenerNotifyLock` may ensure two threads don't call `transportInUse()` at the same time, but that's not actually enough according to the Listener API contract. You actually need mutual exclusion across all Listener methods.